### PR TITLE
Integrated Presence with PubNub stage1

### PR DESCRIPTION
**Pull request by marcin-cebo**

### CODE REVIEW

File: src/main/kotlin/com/pubnub/api/managers/PresenceEventEngineManager.kt [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+import com.pubnub.api.eventengine.EffectDispatcher
+import com.pubnub.api.eventengine.Event
+import com.pubnub.api.eventengine.EventEngineManager
+import com.pubnub.api.eventengine.Sink
+import com.pubnub.api.presence.eventengine.PresenceEventEngine
+import com.pubnub.api.presence.eventengine.effect.PresenceEffectInvocation
+import com.pubnub.api.presence.eventengine.event.PresenceEvent
+
+class PresenceEventEngineManager(
```
Comment (kleewho) [inline src/main/kotlin/com/pubnub/api/managers/PresenceEventEngineManager.kt:11]: This class is the same as the `SubscribeEventEngineManager`. I think we should have just one implementation

File: src/main/kotlin/com/pubnub/api/presence/eventengine/PresenceEventEngine.kt [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import com.pubnub.api.eventengine.Sink
+import com.pubnub.api.eventengine.Source
+import com.pubnub.api.eventengine.transition
+import com.pubnub.api.presence.eventengine.effect.PresenceEffectInvocation
+import com.pubnub.api.presence.eventengine.event.PresenceEvent
+import com.pubnub.api.presence.eventengine.state.PresenceState
+import java.util.concurrent.ExecutorService
+import java.util.concurrent.Executors
+
+class PresenceEventEngine(
```
Comment (kleewho) [inline src/main/kotlin/com/pubnub/api/presence/eventengine/PresenceEventEngine.kt:13]: Generalized EventEngine implementation should be in `c.p.a.eventengine` package and both Subscribe and Presence should use it. Both have exactly the same implementation so there's no need to duplicate it

File: src/main/kotlin/com/pubnub/api/subscribe/Subscribe.kt [hunk 3]
```diff
@@ -6,15 +6,19 @@ import com.pubnub.api.PubNub
...
+import com.pubnub.api.eventengine.Event
+import com.pubnub.api.eventengine.EventEnginesConf
 import com.pubnub.api.managers.ListenerManager
+import com.pubnub.api.managers.PresenceEventEngineManager
+import com.pubnub.api.managers.SubscribeEventEngineManager
+import com.pubnub.api.presence.eventengine.PresenceEventEngine
+import com.pubnub.api.presence.eventengine.effect.PresenceEffectFactory
+import com.pubnub.api.presence.eventengine.effect.effectprovider.HeartbeatProviderImpl
+import com.pubnub.api.presence.eventengine.effect.effectprovider.LeaveProviderImpl
+import com.pubnub.api.presence.eventengine.event.PresenceEvent
```
Comment (kleewho) [inline src/main/kotlin/com/pubnub/api/subscribe/Subscribe.kt:18]: I don't think the package `c.p.a.subscribe` should depend on `c.p.a.presence`. It make sense to use just subscribe without presence. `Subscribe` class should be concerned only with subscribe and presence related things should go into `c.p.a.presence.Presence`

File: src/main/kotlin/com/pubnub/api/subscribe/eventengine/configuration/EventEnginesConfImpl.kt [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+import com.pubnub.api.eventengine.EventEnginesConf
+import com.pubnub.api.eventengine.QueueSinkSource
+import com.pubnub.api.eventengine.Sink
+import com.pubnub.api.eventengine.Source
+import com.pubnub.api.presence.eventengine.effect.PresenceEffectInvocation
+import com.pubnub.api.presence.eventengine.event.PresenceEvent
+import com.pubnub.api.subscribe.eventengine.effect.SubscribeEffectInvocation
+import com.pubnub.api.subscribe.eventengine.event.SubscribeEvent
+
+class EventEnginesConfImpl : EventEnginesConf {
```
Comment (kleewho) [inline src/main/kotlin/com/pubnub/api/subscribe/eventengine/configuration/EventEnginesConfImpl.kt:12]: I have proposal how we can do keep the "configurations" for both presence and subscribe more separate. I will have it in one of the commits

File: src/test/kotlin/com/pubnub/api/subscribe/SubscribeTest.kt [hunk 5]
```diff
@@ -54,6 +61,13 @@ internal class SubscribeTest {
             ),
             subscribeEvent.captured
         )
+
+        assertEquals(PresenceEvent.Joined(setOf(channelToSubscribe), setOf(channelGroupsToSubscribe)), presenceEvent.captured)
```
Comment (kleewho) [inline src/test/kotlin/com/pubnub/api/subscribe/SubscribeTest.kt:65]: None of the Presence things should be tested here

**Review by kleewho — CHANGES_REQUESTED** [review]

**Review by kleewho — APPROVED** [review]
It obviously looks good to me. What do you think?

File: src/test/kotlin/com/pubnub/contract/subscribe/eventEngine/step/EventEngineSteps.kt [hunk 6]
```diff
@@ -118,7 +118,7 @@ class EventEngineSteps(private val state: EventEngineState) {
     fun i_observe_the_following(dataTable: DataTable) {
         await.pollInterval(50, TimeUnit.MILLISECONDS).atMost(500, TimeUnit.MILLISECONDS).untilAsserted {
             val expectedNames = dataTable.asMaps().map { it["type"] to it["name"] }.toList()
-            MatcherAssert.assertThat(state.queuedElements, Matchers.`is`(expectedNames))
+            MatcherAssert.assertThat(state.subscribeQueuedElements, Matchers.`is`(expectedNames))
         }
     }
```
Comment (marcin-cebo) [inline src/test/kotlin/com/pubnub/contract/subscribe/eventEngine/step/EventEngineSteps.kt:124]: I like this refactor. I wish I could write such kotlin code.

File: src/test/kotlin/com/pubnub/api/presence/PresenceTest.kt [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+import org.junit.jupiter.params.provider.MethodSource
+import java.time.Duration
+
+class PresenceTest {
+
+    companion object {
+        @JvmStatic
+        fun eventAndMethodProvider(): List<Arguments> {
+            return listOf(
+                arguments(
```
Comment (marcin-cebo) [inline src/test/kotlin/com/pubnub/api/presence/PresenceTest.kt:26]: I like test.

**Pull request merged**